mv_populating_from_existing_data_during_node_stop_test failed: View ks.tm_table_mv_2 not built

Scylla version 77a0d5c5da46521d70a29df62338ee27f6b5a518
scylladb/scylla-dtest@3d5f8781ba2ce9cd9053cc7d615a8aef695a1972

Seen in [next/668/testReport/materialized_views_test/TestMaterializedViews/mv_populating_from_existing_data_during_node_stop_test](http://jenkins.cloudius-systems.com:8080/view/master/job/scylla-master/job/next/668/testReport/materialized_views_test/TestMaterializedViews/mv_populating_from_existing_data_during_node_stop_test/):
```
Error Message
View ks.tm_table_mv_2 not built

Stacktrace
  File "/usr/lib64/python2.7/unittest/case.py", line 367, in run
    testMethod()
  File "/jenkins/workspace/scylla-master/next/scylla-dtest/materialized_views_test.py", line 587, in mv_populating_from_existing_data_during_node_stop_test
    self._mv_populating_from_existing_data_during_changes_test('stop node', nodes=4, rf=3, mvs=10, prefill=40000, fail=False)
  File "/jenkins/workspace/scylla-master/next/scylla-dtest/materialized_views_test.py", line 653, in _mv_populating_from_existing_data_during_changes_test
    wait_for_view(cluster=self.cluster, session=session, ks=tm.keyspace, view=mv_name)
  File "/jenkins/workspace/scylla-master/next/scylla-dtest/scylla_tools.py", line 905, in wait_for_view
    raise Exception(error_msg)
```

Node logs are [here](http://jenkins.cloudius-systems.com:8080/view/master/job/scylla-master/job/next/668/artifact/logs-release.2/1554102223280_materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_stop_test/)

> By the way - _some_ errors like `std::_Nested_exception<no_such_column_family> (Can't find a column family with UUID ce56a440-a7b3-11e9-b00c-000000000000): std::out_of_range (_Map_base::at)` can still occur in the logs when somebody replaces an old view definition with a new one, because there's a preemption point between removing the view from the database and stopping its view building. View builder is ready for that and these errors are totally benign - they're just last cries of stale build steps before the whole building process is aborted, but nonetheless they may appear.
> 
> Fortunately, these errors are already ignored, because they'll all start with `Error applying view update (...)`, which is on the ignore list. I'm just writing it here so people know what might be the source of these benign errors.

That's true.  Another note for future readers of this issue is that they are ignored only if the test gets to `check_errors`.  unlike cases that call `_mv_populating_from_existing_data` that sets:
https://github.com/scylladb/scylla-dtest/blob/5936f11fb1e3668e836a769306991563d1f1fe29/materialized_views_test.py#L563
`_mv_populating_from_existing_data_during_changes_test` doesn't set this (or other errors) in `self.ignore_log_patterns`.  Ignoring happens as a side effect of calling check_errors in:
https://github.com/scylladb/scylla-dtest/blob/5936f11fb1e3668e836a769306991563d1f1fe29/materialized_views_test.py#L696-L699

For example, see https://jenkins.scylladb.com/view/scylla-3.2/job/scylla-3.2/job/dtest-release/20/testReport/materialized_views_test/TestMaterializedViews/mv_populating_from_existing_data_during_node_restart_test_2/
```
Traceback (most recent call last):
  File "/usr/lib64/python2.7/unittest/case.py", line 396, in run
    self.tearDown()
  File "/jenkins/workspace/scylla-3.2/dtest-release/scylla-dtest/dtest.py", line 907, in tearDown
    self.cleanUpCluster()
  File "/jenkins/workspace/scylla-3.2/dtest-release/scylla-dtest/dtest.py", line 923, in cleanUpCluster
    raise AssertionError('Unexpected error in %s node log: %s' % (node.name, errors))
AssertionError: Unexpected error in node1 node log: ["ERROR 2019-12-09 09:20:48,293 [shard 0] view - Error applying view update to 127.0.82.1: std::_Nested_exception<no_such_column_family> (Can't find a column family with UUID 2f781780-1a65-11ea-aa12-000000000000): std::out_of_range (_Map_base::at)\nWARN  2019-12-09 09:20:48,293 [shard 0] view - Error executing build step for base ks.tm_table: std::_Nested_exception<no_such_column_family> (Can't find a column family with UUID 2f781780-1a65-11ea-aa12-000000000000): std::out_of_range (_Map_base::at)\nERROR 2019-12-09 09:20:48,294 [shard 1] view - Error applying view update to 127.0.82.1: std::_Nested_exception<no_such_column_family> (Can't find a column family with UUID 2f781780-1a65-11ea-aa12-000000000000): std::out_of_range (_Map_base::at)\nWARN  2019-12-09 09:20:48,294 [shard 1] view - Error executing build step for base ks.tm_table: std::_Nested_exception<no_such_column_family> (Can't find a column family with UUID 2f781780-1a65-11ea-aa12-000000000000): std::out_of_range (_Map_base::at)"]
```

This happened after it failed [here](https://jenkins.scylladb.com/view/scylla-3.2/job/scylla-3.2/job/dtest-release/20/testReport/materialized_views_test/TestMaterializedViews/mv_populating_from_existing_data_during_node_restart_test/):
```
Traceback (most recent call last):
  File "/usr/lib64/python2.7/unittest/case.py", line 367, in run
    testMethod()
  File "/jenkins/workspace/scylla-3.2/dtest-release/scylla-dtest/materialized_views_test.py", line 626, in mv_populating_from_existing_data_during_node_restart_test
    self._mv_populating_from_existing_data_during_changes_test('restart node', nodes=4, rf=3, mvs=10, prefill=40000, fail=False)
  File "/jenkins/workspace/scylla-3.2/dtest-release/scylla-dtest/materialized_views_test.py", line 684, in _mv_populating_from_existing_data_during_changes_test
    wait_for_view(cluster=self.cluster, session=session, ks=tm.keyspace, view=mv_name)
  File "/jenkins/workspace/scylla-3.2/dtest-release/scylla-dtest/scylla_tools.py", line 916, in wait_for_view
    raise Exception(error_msg)
Exception: View ks.tm_table_mv_2 not built
```

--- a/db/schema_tables.cc
+++ b/db/schema_tables.cc
@@ -1050,16 +1050,19 @@ static void merge_tables_and_views(distributed<service::storage_proxy>& proxy,
 
             auto& mm = service::get_local_migration_manager();
             auto it = columns_changed.begin();
-            std::vector<future<>> notifications;
-            notifications.reserve(tables_diff.size() + views_diff.size());
-            auto notify = [&] (auto& r, auto&& f) { boost::range::transform(r, std::back_inserter(notifications), f); };
-            notify(tables_diff.created, [&] (auto&& gs) { return mm.notify_create_column_family(gs); });
-            notify(tables_diff.altered, [&] (auto&& gs) { return mm.notify_update_column_family(gs, *it++); });
-            notify(tables_diff.dropped, [&] (auto&& dt) { return mm.notify_drop_column_family(dt.schema); });
-            notify(views_diff.created, [&] (auto&& gs) { return mm.notify_create_view(view_ptr(gs)); });
-            notify(views_diff.altered, [&] (auto&& gs) { return mm.notify_update_view(view_ptr(gs), *it++); });
+            auto notify = [&] (auto& r, auto&& f) {
+                auto notifications = r | boost::adaptors::transformed(f);
+                when_all(notifications.begin(), notifications.end()).get();
+            };
+            // View drops are notified first, because a table can only be dropped if its views are already deleted
             notify(views_diff.dropped, [&] (auto&& dt) { return mm.notify_drop_view(view_ptr(dt.schema)); });
-            when_all(notifications.rbegin(), notifications.rend()).get();
+            notify(tables_diff.dropped, [&] (auto&& dt) { return mm.notify_drop_column_family(dt.schema); });
+            // Table creations are notified first, in case a view is created right after the table
+            notify(tables_diff.created, [&] (auto&& gs) { return mm.notify_create_column_family(gs); });
+            notify(views_diff.created, [&] (auto&& gs) { return mm.notify_create_view(view_ptr(gs)); });
+            // Table altering is notified first, in case new base columns appear
+            notify(tables_diff.altered, [&] (auto&& gs) { return mm.notify_update_column_family(gs, *it++); });
+            notify(views_diff.altered, [&] (auto&& gs) { return mm.notify_update_view(view_ptr(gs), *it++); });
         });
     }).get();
 }

